### ISSUE TITLE
Are checkpoints broken in >= 2.16?

### ISSUE THREAD

**Issue by IvanUkhov** [tensorflow/tensorflow#82209]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

2.16, 2.17

### Custom code

No

### Current behavior?

The example given in https://www.tensorflow.org/guide/checkpoint does not seem to work as expected in 2.16 and 2.17, while working fine in 2.15. After restoring and restarting the training process, it starts training from the very beginning.

### Standalone code to reproduce the issue

```shell
https://colab.research.google.com/drive/1n76Mu5BhdBJBSXc7cXYJMr0lMDER2JRa?usp=sharing
```

**Comment by Venkat6871** [comment]
Hi **@IvanUkhov** ,
Apologies for the delay, and thank you for raising your concern. I tried running your code on Colab using TensorFlow 2.17.0 and noticed some numerical loss. However, when using TensorFlow 2.15.0, I observed that the loss value was NaN.
Could you please confirm if my approach is correct?
For your reference, I have shared the [gist](https://colab.sandbox.google.com/gist/Venkat6871/e25040901a6960adc43906d6b29f4c80/82209_tf-2-17-0-2-15-0.ipynb) here.
Thank you!

**Comment by IvanUkhov** [comment]
Thank you for looking into this!

Well, I simply ran the code from this page https://www.tensorflow.org/guide/checkpoint with three different workbenches under Vertex AI: 2.15, 2.16, and 2.17. I did not observe any NaNs when running. Just that the loss was reset to its original value in version 2.16 and 2.17, indicating that the restoration from the checkpoint did not work out.
